modules/core/src/ecdh.ts

```diff
@@ -0,0 +1,57 @@
+/**
+ * @prettier
+ *
+ * Utility methods for Ellipic-Curve Diffie-Hellman (ECDH) shared secret generation
+ *
+ * > Elliptic-curve Diffie–Hellman (ECDH) is a key agreement protocol that allows two parties, each having an
+ * > elliptic-curve public–private key pair, to establish a shared secret over an insecure channel.
+ * > This shared secret may be directly used as a key, or to derive another key. The key, or the derived key, can then
+ * > be used to encrypt subsequent communications using a symmetric-key cipher. It is a variant of the Diffie–Hellman
+ * > protocol using elliptic-curve cryptography.
+ *
+ * https://en.wikipedia.org/wiki/Elliptic-curve_Diffie%E2%80%93Hellman
+ */
+
+import * as assert from 'assert';
+import * as bip32 from 'bip32';
+import * as secp256k1 from 'secp256k1';
+import * as utxolib from '@bitgo/utxo-lib';
+
+/**
+ * Calculate the Elliptic Curve Diffie Hellman
+ * @param privateKey HDNode of private key
+ * @param publicKey [neutered] HDNode of public key
+ * @returns Buffer public key buffer that can be used as shared secret (see note)
+ */
+export function getSharedSecret(
+  privateKey: bip32.BIP32Interface | utxolib.ECPair | Buffer,
+  publicKey: bip32.BIP32Interface | Buffer
+): Buffer {
+  if (privateKey.constructor.name === 'BIP32') {
+    if (!privateKey.privateKey) {
+      throw new Error(`privateNode must be private key`);
+    }
+    privateKey = privateKey.privateKey;
+  } else if (privateKey instanceof utxolib.ECPair) {
+    privateKey = privateKey.d.toBuffer(32);
+  }
+
+  if (!Buffer.isBuffer(publicKey)) {
+    publicKey = publicKey.publicKey;
+  }
+
+  if (!Buffer.isBuffer(privateKey) || !Buffer.isBuffer(publicKey)) {
+    throw new Error(`invalid state`);
```

Comment:
Maybe add some more detail to this message?

Reply:
This is more to satisfy `tsc` than anything else. You'd really have put in some effort to get here. I would have used `assert` but I didn't want the extra import.

Reply:
Wait a second, I actually do have `assert` already imported. I think for some reason the type assertion signature was not recognized. In any case, the user will almost certainly never see this and us devs have the stack trace.